src/modules/ods/ods.service.test.ts

```diff
@@ -0,0 +1,104 @@
+import { OdsService } from './ods.service';
+import { ODS_ENTITIES, OdsStoredProcedureInput } from './dto/ods-payloads.dto';
+import { CUSTOMERS } from '@ukef/constants';
+import { DataSource, QueryRunner } from 'typeorm';
+
+describe('OdsService', () => {
+  let service: OdsService;
+  let mockQueryRunner: jest.Mocked<QueryRunner>;
+  let mockDataSource: jest.Mocked<DataSource>;
+
+  beforeEach(() => {
+    mockQueryRunner = {
+      query: jest.fn(),
+      release: jest.fn(),
+    } as unknown as jest.Mocked<QueryRunner>;
+
+    mockDataSource = {
+      createQueryRunner: jest.fn().mockReturnValue(mockQueryRunner),
+    } as unknown as jest.Mocked<DataSource>;
+
+    service = new OdsService(mockDataSource, null);
+  });
+
+  describe('callOdsStoredProcedure', () => {
+    it('should call the stored procedure with the query runner', async () => {
+      const mockInput: OdsStoredProcedureInput = service.createOdsStoredProcedureInput(ODS_ENTITIES.CUSTOMER, {
+        customer_party_unique_reference_number: CUSTOMERS.EXAMPLES.PARTYURN,
+      });
+
+      const mockOutputBody = JSON.stringify({ id: '123', name: 'Test Customer' });
+      const mockResult = [{ output_body: mockOutputBody }];
+      mockQueryRunner.query.mockResolvedValue(mockResult);
+
+      const result = await service.callOdsStoredProcedure(mockInput);
+
+      expect(mockDataSource.createQueryRunner).toHaveBeenCalled();
+      expect(mockQueryRunner.query).toHaveBeenCalledWith(
+        expect.stringMatching(
+          /DECLARE @output_body NVARCHAR\(MAX\);\s*EXEC t_apim\.sp_ODS_query @input_body=@0, @output_body=@output_body OUTPUT;\s*SELECT @output_body as output_body\s*/,
+        ),
+        [JSON.stringify(mockInput)],
+      );
+      expect(result).toEqual(mockOutputBody);
+      expect(mockQueryRunner.release).toHaveBeenCalled();
+    });
+  });
+
+  describe('findCustomer', () => {
+    it('should return a customer the customer urn and name when findCustomer is called', async () => {
```

Comment:
Can we add an assertion where ODS does not return any response.
Bonus: connection failure to ODS please.

Reply:
Have added one calling ODS throws which should cover a failed connection or ODS not returning us a response or timing out